x-pack/plugins/maps/server/mvt/get_tile.ts

```diff
@@ -0,0 +1,268 @@
+
+interface ESBounds {
+  top_left: {
+    lon: number;
+    lat: number;
+  };
+  bottom_right: {
+    lon: number;
+    lat: number;
+  };
+}
+
+export async function getTile({
+  logger,
+  callElasticsearch,
+  index,
+  geometryFieldName,
+  x,
+  y,
+  z,
+  requestBody = {},
+}: {
+  x: number;
+  y: number;
+  z: number;
+  geometryFieldName: string;
+  index: string;
+  callElasticsearch: (type: string, ...args: any[]) => Promise<unknown>;
+  logger: Logger;
+  requestBody: any;
+}): Promise<Buffer | null> {
+  const geojsonBbox = tileToGeoJsonPolygon(x, y, z);
+
+  let resultFeatures: Feature[];
+  try {
+    let result;
+    try {
+      const geoShapeFilter = {
+        geo_shape: {
+          [geometryFieldName]: {
+            shape: geojsonBbox,
+            relation: 'INTERSECTS',
+          },
+        },
+      };
+
+      // http://localhost:5601/veq/api/maps/mvt/getTile?x=0&y=0&z=0&geometryFieldName=coordinates&index=world_countries_v1&fields=_id&requestBody=(_source:(excludes:!()),docvalue_fields:!(),query:(bool:(filter:!((match_all:())),must:!(),must_not:!(),should:!())),script_fields:(),size:10000,stored_fields:!(%27*%27))
+
+      requestBody.query.bool.filter.push(geoShapeFilter);
+
+      const esSearchQuery = {
+        index,
+        body: requestBody,
+      };
+
+      const esCountQuery = {
+        index,
+        body: {
+          query: requestBody.query,
+        },
+      };
+
+      const countResult = await callElasticsearch('count', esCountQuery);
+
+      // @ts-expect-error
+      if (countResult.count > requestBody.size) {
+        // Generate "too many features"-bounds
+        const bboxAggName = 'data_bounds';
+        const bboxQuery = {
+          index,
+          body: {
+            size: 0,
+            query: requestBody.query,
+            aggs: {
+              [bboxAggName]: {
+                geo_bounds: {
+                  field: geometryFieldName,
+                },
+              },
+            },
+          },
+        };
+
+        const bboxResult = await callElasticsearch('search', bboxQuery);
+
+        // @ts-expect-error
+        const bboxForData = esBboxToGeoJsonPolygon(bboxResult.aggregations[bboxAggName].bounds);
+
+        resultFeatures = [
+          {
+            type: 'Feature',
+            properties: {
+              [KBN_TOO_MANY_FEATURES_PROPERTY]: true,
+            },
+            geometry: bboxForData,
+          },
+        ];
+      } else {
+        // Perform actual search
+        result = await callElasticsearch('search', esSearchQuery);
+
+        // @ts-expect-error
+        const hitsFeatures: Array<Feature | null> = result.hits.hits.map(
+          (hit: any): Feature | null => {
+            let geomType:
+              | 'Point'
+              | 'MultiPoint'
+              | 'LineString'
+              | 'MultiLineString'
+              | 'Polygon'
+              | 'MultiPolygon';
+            const geometry = hit._source[geometryFieldName];
```

Comment:
What about geometry indexed as WKT or containing an  array of shapes? Maybe this should use `elasticsearch_geo_utils. geoShapeToGeometry` logic?